### ISSUE TITLE
Possible error in `alpha` function

### ISSUE THREAD

**Issue by HalfVoxel** [andywiecko/BurstTriangulator#188]
This method:

```
public readonly float alpha(float D, float d, bool initial) {
    var k = (int)math.round(math.log2(0.5f * d / D));
    var alpha = D / d * (1 << k);
    return initial ? alpha : 1 - alpha;
}
```

is called with
```
D = 0.001 (by default)
d = distance between two endpoints in an segment
```

This means that if `d < 2*D` then `k` could become negative (d = 0.00001, D = 0.001 gives `k = -8`). Trying to bitshift with a negative value produces unexpected results: `1 << (-8) = 16777216`.

The `alpha` value is used as an input to `math.lerp` to calculate a point on the segment, and a value which is not in the range [0..1] will produce a point that is not on the segment.

This could potentially lead to a bug in the program.
I haven't replicated any case where this happens, but it looks like a potential issue.

**Comment by andywiecko** [comment]
@HalfVoxel 
Thanks for reporting this! 

This `(1 << k)` was used for fast $2^k$ calculation, which certainly does not handle when $k<0$.
I added this issue to the project board.
Probably the easiest fix is just to modify this like: 

```csharp
var ... = k < 0 ? math.pow(2, k) : 1 << k;
```

Some tests are required.

**Comment by HalfVoxel** [comment]
Reading the paper it's from, I think the code can additionally be simplified to remove `D` altogether.
The method will return the exact same values for as you scale `D` by a power of two (for example, `D=0.01` gives the exact same result as `D=0.01*2^7 = 1.28`). If I read the code correctly. So one can just pick the arbitrary constant `D=1` and remove one multiplication, and avoid having `D` in the config at all.

The only other purpose I can see to `D` is if they actually mean that you should use `2^max(k, 0)`. Meaning `D` is the minimum concentric circle radius. But that seems kinda weird, and they do say "for all integers i", which implies `i \in Z`.

By the way. In the paper, they always split at the midpoint in the first split. But in your code, you just invert `alpha` (instead of setting alpha=0.5). Is there a reason for this?
![bild](https://github.com/user-attachments/assets/375273cb-e0ea-48af-932a-58621950e3fd)

**Comment by andywiecko** [comment]
Nice observations regarding `D`. Probably could be done something with this in the future!

> By the way. In the paper, they always split at the midpoint in the first split. But in your code, you just invert `alpha` (instead of setting alpha=0.5). Is there a reason for this?

First split is **always** at the midpoint (`alpha = 0.5`) check the code:

![image](https://github.com/user-attachments/assets/98fac8ea-ce53-4d7f-92bb-d257a03a18f5)

The additional switch `i < initialPointsCount` is introduced to provide symmetry in splitting.

**Comment by HalfVoxel** [comment]
> First split is always at the midpoint (alpha = 0.5) check the code:

Ah. I got confused because the last parameter to the `alpha` is named `initial`.

**Comment by andywiecko** [comment]
Hi @HalfVoxel,

I believe that this issue has been resolved with the recent commits.

In 5366e9ba46448ba8154ea6aaa87d6e3f40d3fc02, I added comments regarding the alpha method. I hope this will improve readability for the contributors. I also removed the confusing `initial` parameter and moved this check into the triangulation step. Additionally, in 57f22a536f51a318d98d8d09d10ff1669ade6da2, I fixed the issue for $k < 0$ and added corresponding unit tests.

I believe this should work for now.

Best,
Andrzej
